File: Sources/InterfaceGraphique/Editor/EditorState/OfflineEditorState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using InterfaceGraphique.Entities;

namespace InterfaceGraphique.Editor.EditorState 
{
    public class OfflineEditorState : AbstractEditorState
    {
        public OfflineEditorState()
        {
            this.InitializeCallbacks();
        }

        public override void JoinEdition(MapEntity mapEntity)
        {
            FonctionsNatives.setOnlineClientType((int)OnlineClientType.OFFLINE_EDITION);
            FonctionsNatives.setCurrentPlayerSelectionColorToDefault();
            //FonctionsNatives.clearUsers();

            this.SetCallbacks();
        }

        protected override void SaveMap()
        {
            if (Editeur.mapManager.CurrentMapAlreadySaved())
                Task.Run(() =>
                    Editeur.mapManager.SaveMap()
                    );
        }

        protected override void CurrentUserCreatedPortal(string startUuid, IntPtr startPos, float startRotation, IntPtr startScale, string endUuid, IntPtr endPosition, float endRotation, IntPtr endScale)
        {
            SaveMap();
        }

        protected override void CurrentUserCreatedWall(string uuid, IntPtr pos, float rotation, IntPtr scale)
        {
            SaveMap();
        }

        protected override void CurrentUserCreatedBoost(string uuid, IntPtr startpos, float rotation, IntPtr scale)
        {
            SaveMap();
        }

        protected override void CurrentUserObjectTransformChanged(string uuid, IntPtr pos, float rotation, IntPtr scale)
        {
            this.inTransformation = true;
        }

        protected override void CurrentUserSelectedObject(string uuidselected, bool isSelected, bool deselectAll)
        {
            // Do nothing.
        }

        protected override void CurrentUserChangedControlPoint(string uuid, IntPtr position)
        {
            SaveMap();
        }

        protected override void CurrentUserDeletedNode(string uuid)
        {
            SaveMap();
        }

        public override Task LeaveEdition()
        {
            // Nothing
            return Task.FromResult(0);
        }

        public override void HandleCoefficientChanges(float coefficientFriction, float coefficientAcceleration, float coefficientRebond)
        {
            // Do nothing.
        }
    }
}

File: AirHockeyLightClient/AirHockeyLightClient/Noyau/Editor/States/OfflineEditorState.swift
///////////////////////////////////////////////////////////////////////////////
/// @file OfflineEditorState.swift
/// @author Pierre To
/// @date 2017-11-04
/// @version 1
///
/// @addtogroup log3900 LOG3990
/// @{
///////////////////////////////////////////////////////////////////////////////

import SceneKit

///////////////////////////////////////////////////////////////////////////
/// @class OfflineEditorState
/// @brief Cette classe implémente l'état du mode d'édition hors ligne.
///
/// @author Pierre To
/// @date 2017-11-04
///////////////////////////////////////////////////////////////////////////
class OfflineEditorState: EditorState {
    
    /// Instance singleton
    static let instance = OfflineEditorState()
    
    override func joinEdition(mapEntity: MapEntity) {
        FacadeModele.instance.setCurrentUserDefaultColor()
    }
    
    override func leaveEdition() {}
    override func currentUserCreatedBoost(uuid: String, pos: SCNVector3, rotation: Float, scale: SCNVector3) {}
    override func currentUserCreatedWall(uuid: String, pos: SCNVector3, rotation: Float, scale: SCNVector3) {}
    override func currentUserCreatedPortal(startUuid: String, startPos: SCNVector3, startRotation: Float, startScale: SCNVector3,
                                           endUuid: String, endPos: SCNVector3, endRotation: Float, endScale: SCNVector3) {}
    override func currentUserSelectedObject(uuidSelected: String, isSelected: Bool, deselectAll: Bool) {}
    override func currentUserObjectTransformChanged(uuid: String, pos: SCNVector3, rotation: Float, scale: SCNVector3) {}
    override func currentUserControlPointChanged(uuid: String, pos: SCNVector3) {}
    override func currentUserDeletedNode(uuid: String) {}
    override func currentUserChangedCoefficient(coefficientFriction: Float, coefficientRebond: Float, coefficientAcceleration: Float) {}
}

///////////////////////////////////////////////////////////////////////////////
/// @}
///////////////////////////////////////////////////////////////////////////////
